lib/bap_llvm/llvm_macho_loader.hpp

```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+        return failure("request data beyond object bounds");
+    const char *ptr = prim::get_raw_data(obj) + off;
+    T t;
+    std::copy_n(ptr, sizeof(T), reinterpret_cast<char *>(&t));
+    return success(t);
+}
+
+// Please, pay attention here, that it's not a good idea try to use a relocation_info
+// struct directly. We should rely on any_relocation_info instead. Llvm guys actually provide
+// a very useful functions that allow us not to touch all bits fields from relocation_info struct,
+// because they become absolutely messy when we have to take into account endianess, e.g.
+// when we have a deal with big endian objects on a little endian host.
+//
+// Example:
+// We would like to find out what is a r_symbolnum value from relocation_info struct.
+// And we know that this value is in first 24 bits of 32-bit value.
+// So, what do we have:
+// 00 00 01 40 - uint32_t in BE (big endian) file
+// 40 01 00 00 - uint32_t that we read on EL (little endian) host
+//
+// As one can notice from BE representation, a correct value of r_symbolnum should be just 1 (00 00 01).
+// And we are on a EL host, so we don't have any other options then to read values as a little endian.
+// And here we go:
+// 1) if we read 24 bits of 40 01 00 00 we will get 00 01 40
+// 2) if we swap this value again and get 00 00 01 40, and read first 24 bits, we will get 01 00 00.
+// Neither of this cases is correct.
+// So we need to play with shifts and masks here, what is actually done in llvm functions
+// like getPlainRelocationSymbolNum
+error_or<MachO::any_relocation_info> get_rel(const macho &obj, uint32_t off) {
+    auto any = get_data<MachO::any_relocation_info>(obj, off);
+    if (!any) return any;
+    if (obj.isLittleEndian() != sys::IsLittleEndianHost)
+        MachO::swapStruct(*any);
+    if (!obj.is64Bit() && obj.isRelocationScattered(*any)) {
+        return failure("scattered relocations are not supported");
+    }
+    return any;
+}
+
+// Size in bytes of relocation entry. It coulbe either size of
+// relocation_info, scattered_relocation_info, any_relocation_info
+#define RELOCATION_ENTRY_SIZE 8
+
+// Note, that if r_extern is set to 1, then r_symbolnum contains an index in symbtab, and
+// section index otherwise.
+void iterate_dyn_relocations(const macho &obj, uint32_t off, uint32_t num, ogre_doc &s) {
     for (std::size_t i = 0; i < num; ++i) {
-        const MachO::relocation_info *rel = reinterpret_cast<const MachO::relocation_info *>(ptr + next);
-        if (rel->r_address & MachO::R_SCATTERED) {
-            // not supported
-            next = next + sizeof(MachO::scattered_relocation_info);
-        } else {
-            if (rel->r_extern)
-                symbol_reference(obj, rel->r_symbolnum, rel->r_address, s);
-            next = next + sizeof(MachO::relocation_info);
-        }
+        auto rel = get_rel(obj, off + i * RELOCATION_ENTRY_SIZE);
+        if (rel && obj.getPlainRelocationExternal(*rel))
```

Comment:
the `getPlainRelocationExternal` function depends on the endiannes, however, you already swapped the relocation bytes. Do you think that it is correct? I would suggest you to re-implement those LLVM functions and do not rely on them. Looks like this part of LLVM is buggy and was never tested, probably it is just easier to reimplement this stuff here.